ikarus/finiteelements/fehelper.hh

```diff
@@ -38,4 +38,80 @@ auto localSolutionBlockVector(const typename Traits::FERequirementType::Solution
         localX[i][j] = x[localView.index(localView.tree().child(j).localIndex(i))[0]];
   return localX;
 }
+
+/**
+ * \brief A function which loops over all the nodes of a tree and performs different actions
+ * for a power node and a leaf node depending on the corresponding functor passed.
+ *
+ * \details This function is inspired from the function Dune::TypeTree::Detail::forEachNode
+ * available in dune/typetree/traversal.hh
+ *
+ * \tparam T Type of tree.
+ * \tparam ChildIndex Type of the index of the child.
+ * \tparam PowerFunc Type of the functor called for a power node.
+ * \tparam LeafFunc Type of the functor called for a leaf node.
+ * \param tree Tree whose nodes are to be looped over.
+ * \param childIndex The index of the child.
+ * \param powerFunc A functor to be called for a power node.
+ * \param leafFunc A functor to be called for a leaf node.
+ */
+template <class T, class ChildIndex, class PowerFunc, class LeafFunc>
+void forEachLeafOrPowerNode(T&& tree, ChildIndex&& childIndex, PowerFunc&& powerFunc, LeafFunc&& leafFunc) {
+  using Tree = std::decay_t<T>;
+  if constexpr (Tree::isLeaf) {
+    leafFunc(tree);
+  } else if constexpr (Tree::isPower) {
+    if constexpr (Tree::template Child<Dune::Indices::_0>::Type::isComposite) {
+      auto indices = std::make_index_sequence<Tree::degree()>{};
+      Dune::Hybrid::forEach(indices, [&](auto i) { forEachLeafOrPowerNode(tree.child(i), i, powerFunc, leafFunc); });
+    } else
+      powerFunc(tree, childIndex);
+  } else {
+    auto indices = std::make_index_sequence<Tree::degree()>{};
+    Dune::Hybrid::forEach(indices, [&](auto i) { forEachLeafOrPowerNode(tree.child(i), i, powerFunc, leafFunc); });
+  }
+}
+
+/**
+ * \brief Get the global flat indices for the provided basis.
+ *
+ * \details The global indices are collected in a FlatInterLeaved order. I.e. x_0, y_0, z_0, ..., x_n, y_n, z_n
+ * This function can handle a scalar basis, power basis, and a composite basis.
+ *
+ * \tparam LocalView Type of the local view
+ *
+ * \param localView Local view of the element.
+ * \param globalIndices Output vector to store global indices.
+ */
+template <typename LocalView>
+void globalFlatIndices(const LocalView& localView, std::vector<typename LocalView::MultiIndex>& globalIndices) {
```

Comment:
And the namespace name FEHelper is also very generic. Maybe without this additional namespace?

Reply:
Passing the finite element is may be better I guess and then the local view can be accessed simply via `fe.localView()`.
Since it helps the finite elements to manage its global indices, may be the namespace is okay, but I do the following from the other comment?
```cpp
using FEHelper::globalFlatIndices;
globalFlatIndices(fe.localView(), dofs);
```

Reply:
Not sure what your question is actually at the end here. We can talk about it later.